_posts/2020-04-27-ensuring-resolvers-aren't-rejected.md

```diff
@@ -0,0 +1,79 @@
+
+Below is a trivial example where I use a VTL template to create a JSON document that will perform a DynamoDB PutItem operation. There is no syntax highlighting support for VTL in markdown so I have included an image instead.
+
+![vtl template]({{ site.url }}/img/createThingVtlTemplate.png)
+
+The test will look something like the following.
+
+```typescript
+import * as utils from 'amplify-appsync-simulator/lib/velocity/util'
+import { map } from 'amplify-appsync-simulator/lib/velocity/value-mapper/mapper'
+import { Compile, parse } from 'amplify-velocity-template'
+import * as fs from 'fs'
+import * as path from 'path'
+import { expect } from 'chai'
+
+describe('Create Thing', () => {
+  it('should create the correct id value and the DynamoDB data type descriptor object ', () => {
+    const templatePath = path.resolve(__dirname, '../templates/createSMP-request.vtl')
+    const createSMPTemplate = fs.readFileSync(templatePath, { encoding: 'utf8' })
+    const ast = parse(createSMPTemplate)
+    const compiler = new Compile(ast, {
+      valueMapper: map,
+      escape: false
+    })
+    const args = {
+      thing: {
+        attribute1: '000',
+        attribute2: '123',
+        attribute3: '456'
+      }
+    }
+    const context = createVtlContext(args)
+    const result = JSON.parse(compiler.render(context))
+    expect(result.key.id.S.startsWith(`thing_${args.thing.attribute1}`)).to.be.true
+    const uuidRegex = /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/
+    expect(result.key.id.S).to.match(uuidRegex)
+  })
+})
+
+function createVtlContext<T> (args: T) {
+  const util = utils.create([], new Date(Date.now()), Object())
+  const context = {
+    args,
+    arguments: args
+  }
+  return {
+    util,
+    utils: util,
+    ctx: context,
+    context
+  }
+}
+```
+
+Having the ability to unit test my resolver mapping templates means I can use these tests to aid my development efforts as well as run these tests as part of a test suite in my CI/CD pipeline, making sure that these templates at least make sense before deploying new resolvers to my GraphQL API.
+
+Although you may have thought it previously impossible to test VTL templates, by following a similar approach to the TypeScript snippet above, you can easily build concise tests that prove the more advanced parts of your VTL templates really do work the way you think they will. By having this capability, you will now be able to experiment with your VTL whilst maintaining confidence in the more intricate areas of what you're building. These templates are the gateway into your API and these methods offer you an opportunity to build the right thing, the first time around.
```

Comment:
Nice.. maybe add a hook like Tim V has in his latest blog 

> Manual tests taking up all your time and not finding issues before they get to users? We can help! 
> Contact Mechanical Rock to Get Started!

Reply:
Would something like "Feel free to get in touch on LinkedIn or Twitter if you have any more questions" suffice?

Reply:
yep - just provide those links.